### PR TITLE
Make Friend join notification chat module

### PR DESCRIPTION
**Pull request by ConorGriffin37**
Off by default for now, but this seems like functionality a lot of game developers have implemented themselves into their games so we should have a setting to make this easier.

### CODE REVIEW

File: CoreScriptsRoot/Modules/Server/ClientChat/DefaultClientChatModules/ChatSettings.lua [hunk 1]
```diff
@@ -19,6 +19,8 @@ module.ShowUserOwnFilteredMessage = true	--Show a user the filtered version of t
 module.ChatOnWithTopBarOff = false
 module.ScreenGuiDisplayOrder = 6 -- The DisplayOrder value for the ScreenGui containing the chat.
 
+module.ShowFriendJoinNotification = false -- Show a notification in the chat when a players friend joins the game.
```
Comment (ConorGriffin37) [inline CoreScriptsRoot/Modules/Server/ClientChat/DefaultClientChatModules/ChatSettings.lua]: TODO: Discuss with product if we might want to have this on by default.

**Comment by spotco** [comment]
i like this 30 minute turnaround

File: CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	end
+
+	if ShowFriendJoinNotification() then
+		Players.PlayerAdded:connect(function(player)
+			local possibleFriends = Players:GetPlayers()
+			for i = 1, #possibleFriends do
+				if player ~= possibleFriends[i] then
+					coroutine.wrap(function()
+						if possibleFriends[i]:IsFriendsWith(player.UserId) then
+							SendFriendJoinNotification(possibleFriends[i], player)
```
Comment (DarraghGriffin) [inline CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua]: Try not to use so many levels of indentation here, use functions!

**Comment by stillunt1tled** [comment]
Although this entirely unnecessary, it would be nice to get a chat notification if one of your friends joined the game.

File: CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+	local function SendFriendJoinNotification(player, joinedFriend)
+		local speakerObj = ChatService:GetSpeaker(player.Name)
+		if speakerObj then
+			speakerObj:SendSystemMessage(string.format("Your friend %s has joined the game.", joinedFriend.Name), "System", FriendMessageExtraData)
+		end
+	end
+
+	if ShowFriendJoinNotification() then
+		Players.PlayerAdded:connect(function(player)
```
Comment (stillunt1tled) [inline CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua]: Why are you scanning every player for friends? I do believe that you could do this instead:
```lua
Players.PlayerAdded:connect(function(player)
     if player:IsFriendsWith(Players.LocalPlayer.UserId) then
          SendFriendJoinNotification(player, Players.LocalPlayer)
     end
end

File: CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua (hunk 2 above)
Comment (stillunt1tled) [inline CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua]: Why are you scanning every player for friends, I get when the script is first started, but every time a player joins? I do believe that you could do this instead:
```lua
Players.PlayerAdded:connect(function(player)
     if player:IsFriendsWith(Players.LocalPlayer.UserId) then
          SendFriendJoinNotification(player, Players.LocalPlayer)
     end
end

File: CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+			local possibleFriends = Players:GetPlayers()
+			for i = 1, #possibleFriends do
+				if player ~= possibleFriends[i] then
```
Comment (stillunt1tled) [inline CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua]: Why are you scanning every player for friends, I get when the script is first started, but every time a player joins? I do believe that you could do this instead:
```lua
Players.PlayerAdded:connect(function(player)
     if player:IsFriendsWith(Players.LocalPlayer.UserId) then
          SendFriendJoinNotification(player, Players.LocalPlayer)
     end
end

Comment (ConorGriffin37) [inline CoreScriptsRoot/Modules/Server/ServerChat/DefaultChatModules/FriendJoinNotifier.lua]: This is a server side script so there is no LocalPlayer.

**Comment by stillunt1tled** [comment]
@ConorGriffin37: This pull is creating issues, see issue #1000. `SendFriendJoinNotification()` and `TrySendFriendNotification()` error with the following message: "attempt to call a string value"

**Pull request merged**